common/src/main/java/com/viaversion/viarewind/protocol/v1_8to1_7_6_10/storage/EntityTracker1_8.java

```diff
@@ -42,166 +42,173 @@
-		} else {
-			vehicles.put(vehicleId, passengerId);
-		}
-	}
-
-	protected void attachEntity(final int target) {
-		try {
-			final PacketWrapper attachEntity = PacketWrapper.create(ClientboundPackets1_8.SET_ENTITY_LINK, user());
-			attachEntity.write(Types.INT, this.clientEntityId()); // vehicle id
-			attachEntity.write(Types.INT, target); // passenger id
-			attachEntity.write(Types.BOOLEAN, false); // leash
-
-			attachEntity.scheduleSend(Protocol1_8To1_7_6_10.class);
-		} catch (Exception e) {
-			ViaRewind.getPlatform().getLogger().log(Level.SEVERE, "Failed to send attach packet", e);
-		}
-	}
-
-	public void setSpectating(int spectating) {
-		if (spectating != this.clientEntityId() && getPassenger(spectating) != -1) {
-			startSneaking();
-			setSpectating(this.clientEntityId());
-			return;
-		}
-		if (this.spectatingClientEntityId != spectating && this.spectatingClientEntityId != this.clientEntityId()) {
-			attachEntity(-1);
-		}
-		this.spectatingClientEntityId = spectating;
-		if (spectating != this.clientEntityId()) {
-			attachEntity(this.spectatingClientEntityId);
-		}
-	}
-
-	public Int2ObjectMap<VirtualHologramEntity> getHolograms() {
-		return holograms;
-	}
-
-	public boolean isSpectator() {
-		return clientEntityGameMode == 3;
-	}
-
-	public void setClientEntityGameMode(int clientEntityGameMode) {
-		this.clientEntityGameMode = clientEntityGameMode;
-	}
-
-	public void updateEntityData(List<EntityData> entityData) {
-		this.entityData.removeIf(first -> entityData.stream().anyMatch(second -> first.id() == second.id()));
-		for (final EntityData data : entityData) {
-			final Object value = data.value();
-			if (value instanceof Item item) {
-				this.entityData.add(new EntityData(data.id(), data.dataType(), item.copy()));
-			} else {
-				this.entityData.add(new EntityData(data.id(), data.dataType(), value));
-			}
-		}
-	}
-
-	public List<EntityData> getEntityData() {
-		return entityData;
-	}
+    private final Int2ObjectMap<VirtualHologramEntity> holograms = new Int2ObjectArrayMap<>();
+    private final Int2IntMap vehicles = new Int2IntArrayMap();
+    private final Int2ObjectMap<UUID> entityIdToUUID = new Int2ObjectArrayMap<>();
+    private final Object2IntMap<UUID> entityUUIDToId = new Object2IntOpenHashMap<>();
+
+    private final List<EntityData> entityData = new ArrayList<>();
+
+    public Integer spectatingClientEntityId;
+    private int clientEntityGameMode;
+
+    public EntityTracker1_8(UserConnection connection) {
+        super(connection, EntityTypes1_8.EntityType.PLAYER);
+    }
+
+    @Override
+    public void addEntity(int id, EntityType type) {
+        super.addEntity(id, type);
+        if (type == EntityTypes1_8.EntityType.ARMOR_STAND) {
+            holograms.put(id, new VirtualHologramEntity(user(), id));
+        }
+    }
+
+    @Override
+    public void removeEntity(int entityId) {
+        super.removeEntity(entityId);
+
+        if (holograms.containsKey(entityId)) {
```

Comment:
Instead of containsKey, always do get() and null check:

VirtualHologramEntity hologram = holograms.get(entityId);
if (hologram != null) {
  // ....
}